lib/reports/dynamic.rb

```diff
@@ -0,0 +1,266 @@
+
+require "services"
+require "clusterable/commitment"
+require "clusterable/holding"
+require "clusterable/ht_item"
+require "clusterable/ocn_resolution"
+
+# This is an attempt at a piece of reusable dynamic reporting
+# scaffolding, with which one can get basic reports like
+# "all holdings by colby, with ocn and status" or
+# "all commitments by arizona where condition is excellent"
+
+module Reports
+  class Dynamic
+    attr_reader :base, :decorations, :restrictions
+    def initialize(base: "", decorations: [], restrictions: [])
+      log "new dynamic report!"
+      @base = validate_clusterable(base)
+      @decorations = add_decorations(decorations)
+      @restrictions = add_restrictions(restrictions)
+    end
+
+    # a clusterable (e.g. @base) value must match a key in model.
+    def validate_clusterable(clusterable)
+      log "validate_clusterable(#{clusterable})"
+      unless Dynamic.model.key?(clusterable)
+        mk = Dynamic.model.keys.join(", ")
+        raise ArgumentError, "clusterable not allowed: #{clusterable}. Must be one of #{mk}."
+      end
+      clusterable
+    end
+
+    # Decorations come in as an array of strings.
+    # The strings should be "x.y" where x=on and y=key,
+    # such that on is a model class, and key a field in that class.
+    def add_decorations(decorations)
+      log "add_decorations(#{decorations.inspect})"
+
+      if decorations.empty?
+        raise ArgumentError, "1+ decorations required, none given"
+      end
+
+      decoration_objects = []
+      decorations.each do |d|
+        unless d.is_a? String
+          raise ArgumentError, "decoration input must be a String"
+        end
+
+        on, key = d.split(".")
+        if on.nil? || key.nil?
+          raise ArgumentError, "decoration (#{d}) must match <on>.<key>"
+        end
+
+        decoration_objects << Reports::Decoration.new(on: on, key: key)
+      end
+
+      if decoration_objects.empty?
+        raise ArgumentError, "1+ decorations required, none passed validation"
+      end
+
+      decoration_objects
+    end
+
+    # Restrictions, unlike decorations, can be empty.
+    # Turn input hashes, like {"holdings.organization" => "umich"}
+    # into Dynamic::Restriction objects:
+    # <Restriction, @on:"holdings", @key:"organization", @val:"umich", ...>
+    def add_restrictions(restrictions)
+      log "add_restrictions(#{restrictions.inspect})"
+      restriction_objects = []
+      restrictions.each do |r|
+        unless r.is_a? Enumerable
+          raise ArgumentError, "restriction input must be an Enumerable"
+        end
+
+        unless r.flatten.size == 2
+          raise ArgumentError, "restriction input must be 2 elements long"
+        end
+
+        on, key = r.flatten.first.split(".")
+        val = r.flatten.last
+        restriction_objects << Reports::Restriction.new(on: on, key: key, val: val)
+      end
+      restriction_objects
+    end
+
+    # Puts it all together.
+    def run
```

Comment:
Doesn't look like this is currently covered, and I would think we would want an integration test that puts this all together

Reply:
Agree, that remains.

Reply:
Right. I have a test in the test suite that says it tests `run`, but then it actually only tests `records` which is a hefty part of `run`, but isn't === `run`.

Reply:
Covered now.